gtda/homology/point_clouds.py

```diff
@@ -457,3 +457,194 @@ def transform(self, X, y=None):
+
+        Returns
+        -------
+        self : object
+
+        """
+        if self.infinity_values is None:
+            self.infinity_values_ = self.max_edge_length
+        else:
+            self.infinity_values_ = self.infinity_values
+
+        self._homology_dimensions = sorted(self.homology_dimensions)
+
+        validate_params({**self.get_params(),
+                         'infinity_values_': self.infinity_values_,
+                         '_homology_dimensions': self._homology_dimensions},
+                        self._hyperparameters)
+        check_array(X, allow_nd=True)
+
+        self._max_homology_dimension = self._homology_dimensions[-1]
+        return self
+
+    def transform(self, X, y=None):
+        """For each point cloud in `X`, compute the relevant persistence
+        diagram as an array of triples [b, d, q]. Each triple represents a
+        persistent topological feature in dimension q (belonging to
+        `homology_dimensions`) which is born at b and dies at d. Only triples
+        in which b < d are meaningful. Triples in which b and d are equal
+        ("diagonal elements") may be artificially introduced during the
+        computation for padding purposes, since the number of non-trivial
+        persistent topological features is typically not constant across
+        samples. They carry no information and hence should be effectively
+        ignored by any further computation.
+
+        Parameters
+        ----------
+        X : ndarray of shape (n_samples, n_points, n_dimensions)
+            Input data. Each entry along axis 0 is a point cloud of
+            ``n_points`` row vectors in ``n_dimensions``-dimensional space.
+
+        y : None
+            There is no need for a target in a transformer, yet the pipeline
+            API requires this parameter.
+
+        Returns
+        -------
+        Xt : ndarray of shape (n_samples, n_features, 3)
+            Array of persistence diagrams computed from the feature arrays in
+            `X`. ``n_features`` equals :math:`\\sum_q n_q`, where :math:`n_q`
+            is the maximum number of topological features in dimension
+            :math:`q` across all samples in `X`.
+        """
+        check_is_fitted(self)
+        X = check_array(X, allow_nd=True)
+
+        Xt = Parallel(n_jobs=self.n_jobs)(
+            delayed(self._gudhi_diagram)(X[i, :, :]) for i in range(
+                X.shape[0]))
+
+        max_n_points = {
+            dim: max(1, np.max([Xt[i][dim].shape[0] for i in range(len(
+                Xt))])) for dim in self.homology_dimensions}
+        min_values = {
+            dim: min([np.min(Xt[i][dim][:, 0]) if Xt[i][dim].size else
+                      np.inf for i in range(len(Xt))]) for dim in
+            self.homology_dimensions}
+        min_values = {
+            dim: min_values[dim] if min_values[dim] != np.inf else 0 for dim
+            in self.homology_dimensions}
```

Comment:
It looks like `VietorisRipsPersistence` and `SparseRips` also share the same logic. Can it be refactored (f.ex. the logic moved to `gtda.homology._utils`?

Reply:
Yes, it can indeed.

Reply:
Good point.